### PR TITLE
fix: check the correctness of the leaf ranges

### PR DESCRIPTION
**Pull request by rach-id**
## Overview

Closes: https://github.com/celestiaorg/nmt/issues/267

## Summary by CodeRabbit

- **New Features**
	- Strengthened error handling for subtree root verification and leaf range calculations to improve robustness against invalid inputs.

- **Bug Fixes**
	- Enhanced validation processes to prevent incorrect operations and ensure reliable range processing.

- **Tests**
	- Refined test cases to focus on relevant scenarios and added new tests for improved coverage of edge cases related to error handling.

### CODE REVIEW

File: proof.go [hunk 1]
```diff
@@ -528,6 +528,10 @@ func (proof Proof) VerifySubtreeRootInclusion(nth *NmtHasher, subtreeRoots [][]b
 			return popIfNonEmpty(&subtreeRoots), nil
 		}
 
+		if end-start == 1 {
```
Comment (rootulp) [inline proof.go:531]: [question] what is the intuition around end-start == 1? If end is non-inclusive and start is inclusive and I want to refer to a single node (i.e. node at index 3) then I assume start=3, end=4. In this case end-start does = 1.

**Review by rootulp — APPROVED** [review]
One question

File: proof.go (hunk 1 above)
Comment (rach-id) [reply proof.go:531]: I guess I need to add a comment there to explain it more. 
At that level, we reached a leaf, and we looked in the subtree roots whether we have it, and also looked in the side nodes if we have it, but we can't find any node. This means that since the begining, the provided ranges do not reference valid inner nodes

File: proof_test.go [hunk 2]
```diff
@@ -1828,3 +1839,31 @@ func TestVerifySubtreeRootInclusion(t *testing.T) {
...
+
+	proof, err := tree.ProveRange(19, 64)
+	require.NoError(t, err)
+
+	require.NotPanics(t, func() {
+		// This hits:
+		// runtime: goroutine stack exceeds 1000000000-byte limit
+		// runtime: sp=0x14020160480 stack=[0x14020160000, 0x14040160000]
+		// fatal error: stack overflow
+		_, err = proof.VerifySubtreeRootInclusion(hasher, subtreeRoots, subtreeWidth, root)
```
Comment (cmwaters) [inline proof_test.go:1866]: Huh? Is this supposed to overflow?

Comment (rootulp) [reply proof_test.go:1866]: it's not supposed to overflow. The comment for the panic was from before the fix, maybe it can be updated to:

```
		// This previously hit:
		// runtime: goroutine stack exceeds 1000000000-byte limit
		// runtime: sp=0x14020160480 stack=[0x14020160000, 0x14040160000]
		// fatal error: stack overflow
		_, err = proof.VerifySubtreeRootInclusion(hasher, subtreeRoots, subtreeWidth, root)
```

now the test verifies that it doesn't panic but causes an error instead

File: proof.go (hunk 1 above)
Comment (rootulp) [reply proof.go:531]: ok thanks!

File: proof_test.go (hunk 2 above)
Comment (rach-id) [reply proof_test.go:1866]:
```suggestion
		// This previously hits:
		// runtime: goroutine stack exceeds 1000000000-byte limit
		// runtime: sp=0x14020160480 stack=[0x14020160000, 0x14040160000]
		// fatal error: stack overflow
		_, err = proof.VerifySubtreeRootInclusion(hasher, subtreeRoots, subtreeWidth, root)
```

**Review by rootulp — APPROVED** [review]

File: proof.go (hunk 1 above)
Comment (cmwaters) [reply proof.go:531]: It would be nice if a comment was added here to explain the logic

**Review by cmwaters — APPROVED** [review]

File: proof.go (hunk 1 above)
Comment (rach-id) [reply proof.go:531]: https://github.com/celestiaorg/nmt/pull/272
